src/goto-symex/symex_main.cpp

```diff
@@ -67,17 +67,65 @@ void symex_transition(
     // 1. the transition from state.source.pc to "to" is not a backwards goto
     // or
     // 2. we are arriving from an outer loop
+
+    // TODO: This should all be replaced by natural loop analysis.
     framet &frame = state.call_stack().top();
     const goto_programt::instructiont &instruction=*to;
     for(const auto &i_e : instruction.incoming_edges)
+    {
       if(i_e->is_goto() && i_e->is_backwards_goto() &&
          (!is_backwards_goto ||
           state.source.pc->location_number>i_e->location_number))
       {
         const auto loop_id =
           goto_programt::loop_id(state.source.function_id, *i_e);
-        frame.loop_iterations[loop_id].count = 0;
+        auto &current_loop_info = frame.loop_iterations[loop_id];
+        current_loop_info.count = 0;
+        current_loop_info.is_active = true;
+
+        // Work out what parent this loop has.
+        for(auto &loop : frame.loop_iterations)
+        {
+          if(
+            loop.second.is_active && !loop.second.is_recursion &&
+            current_loop_info.nesting_level < loop.second.nesting_level)
+            current_loop_info.set_parent(&loop.second);
+        }
       }
+    }
+
+    // And conversely, if we're transitioning out of a loop make sure
```

Comment:
Note a single GOTO could break out of more than one loop at a time, and might be taken unconditionally -- I don't think that case is handled.